internal/driver/subscriptionlistener.go

```diff
@@ -0,0 +1,193 @@
+		policy     = d.serviceConfig.OPCUAServer.Policy
+		mode       = d.serviceConfig.OPCUAServer.Mode
+		certFile   = d.serviceConfig.OPCUAServer.CertFile
+		keyFile    = d.serviceConfig.OPCUAServer.KeyFile
+		resources  = d.serviceConfig.OPCUAServer.Writable.Resources
+	)
+
+	// No need to start a subscription if there are no resources to monitor
+	if len(resources) == 0 {
+		d.Logger.Info("[Incoming listener] No resources defined to generate subscriptions.")
+		return nil
+	}
+
+	// Create a cancelable context for Writable configuration
+	ctxBg := context.Background()
+	ctx, cancel := context.WithCancel(ctxBg)
+	d.ctxCancel = cancel
+
+	ds := service.RunningService()
+	if ds == nil {
+		return fmt.Errorf("[Incoming listener] unable to get running device service")
+	}
+
+	device, err := ds.GetDeviceByName(deviceName)
+	if err != nil {
+		return err
+	}
+	endpoint, err := config.FetchEndpoint(device.Protocols)
+	if err != nil {
+		return err
+	}
+
+	endpoints, err := opcua.GetEndpoints(endpoint)
+	if err != nil {
+		return err
+	}
+	ep := opcua.SelectEndpoint(endpoints, policy, ua.MessageSecurityModeFromString(mode))
+	if ep == nil {
+		return fmt.Errorf("[Incoming listener] Failed to find suitable endpoint")
+	}
+	ep.EndpointURL = endpoint
+
+	opts := []opcua.Option{
+		opcua.SecurityPolicy(policy),
+		opcua.SecurityModeString(mode),
+		opcua.CertificateFile(certFile),
+		opcua.PrivateKeyFile(keyFile),
+		opcua.AuthAnonymous(),
+		opcua.SecurityFromEndpoint(ep, ua.UserTokenTypeAnonymous),
+	}
+
+	client := opcua.NewClient(ep.EndpointURL, opts...)
+	if err := client.Connect(ctx); err != nil {
+		d.Logger.Warnf("[Incoming listener] Failed to connect OPCUA client, %s", err)
+		return err
+	}
+	defer client.Close()
+
+	sub, err := client.Subscribe(
+		&opcua.SubscriptionParameters{
+			Interval: time.Duration(500) * time.Millisecond,
+		}, make(chan *opcua.PublishNotificationData))
+	if err != nil {
+		return err
+	}
+	defer sub.Cancel()
+
+	for i, node := range strings.Split(resources, ",") {
+		deviceResource, ok := ds.DeviceResource(deviceName, node)
+		if !ok {
+			return fmt.Errorf("[Incoming listener] Unable to find device resource with name %s", node)
+		}
+
+		opcuaNodeID, err := buildNodeID(deviceResource.Attributes, SYMBOL)
+		if err != nil {
+			return err
+		}
+
+		id, err := ua.ParseNodeID(opcuaNodeID)
+		if err != nil {
+			return err
+		}
+
+		// arbitrary client handle for the monitoring item
+		handle := uint32(i + 42)
+		// map the client handle so we know what the value returned represents
+		d.resourceMap.mu.Lock()
+		d.resourceMap.res[handle] = node
+		d.resourceMap.mu.Unlock()
+		miCreateRequest := opcua.NewMonitoredItemCreateRequestWithDefaults(id, ua.AttributeIDValue, handle)
+		res, err := sub.Monitor(ua.TimestampsToReturnBoth, miCreateRequest)
+		if err != nil || res.Results[0].StatusCode != ua.StatusOK {
+			return err
+		}
+
+		d.Logger.Infof("[Incoming listener] Start incoming data listening for %s.", node)
+	}
```

Comment:
extract this for loop to another func because the lock and unlock should be outside of the for loop
```
func xxx() {
  lock()
  for {
    ...
  }
  defer unlock()
}
```